workspaces/src/network/mod.rs

```diff
@@ -0,0 +1,169 @@
+
+    // TODO: change return type to Url instead of String
+    /// Rpc endpoint to point our client to
+    fn rpc_url(&self) -> String;
+
+    // TODO: not everything has a helper url. maybe make this optional or remove it into a seprate trait
+    /// The helper URL to create top level account out of for certain networks.
+    fn helper_url(&self) -> String;
+}
+
+#[async_trait]
+pub trait TopLevelAccountCreator {
+    async fn create_tla(
+        &self,
+        id: AccountId,
+        signer: InMemorySigner,
+    ) -> anyhow::Result<CallExecution<Account>>;
+
+    async fn create_tla_and_deploy<P: AsRef<Path> + Send + Sync>(
+        &self,
+        id: AccountId,
+        signer: InMemorySigner,
+        wasm: P,
+    ) -> anyhow::Result<CallExecution<Contract>>;
+}
+
+// NOTE: Not all networks/runtimes will have the ability to be able to do dev_deploy.
+// This trait acts as segmented boundary for only specific networks such as sandbox and testnet.
+pub trait AllowDevAccountCreation {}
+
+#[async_trait]
+pub trait DevAccountDeployer {
+    fn dev_generate(&self) -> (AccountId, InMemorySigner);
+    async fn dev_create(&self) -> anyhow::Result<Account>;
+    async fn dev_deploy<P: AsRef<Path> + Send + Sync>(&self, wasm: P) -> anyhow::Result<Contract>;
+}
+
+#[async_trait]
+impl<T> DevAccountDeployer for T
+where
+    T: TopLevelAccountCreator + NetworkInfo + AllowDevAccountCreation + Send + Sync,
+{
+    fn dev_generate(&self) -> (AccountId, InMemorySigner) {
+        let account_id = crate::rpc::tool::random_account_id();
+        let signer =
+            InMemorySigner::from_seed(account_id.clone(), KeyType::ED25519, DEV_ACCOUNT_SEED);
+
+        let mut savepath = self.keystore_path();
+
+        // TODO: potentially make this into the async version:
+        std::fs::create_dir_all(savepath.clone()).unwrap();
+
+        savepath = savepath.join(account_id.to_string());
+        savepath.set_extension("json");
+        signer.write_to_file(&savepath);
+
+        (account_id, signer)
+    }
+
+    async fn dev_create(&self) -> anyhow::Result<Account> {
+        let (account_id, signer) = self.dev_generate();
+        let account = self.create_tla(account_id.clone(), signer).await?;
+        account.into()
+    }
+
+    async fn dev_deploy<P: AsRef<Path> + Send + Sync>(&self, wasm: P) -> anyhow::Result<Contract> {
+        let (account_id, signer) = self.dev_generate();
+        let contract = self
+            .create_tla_and_deploy(account_id.clone(), signer, wasm)
+            .await?;
+        contract.into()
+    }
+}
+
+pub trait AllowStatePatching {}
+
+#[async_trait]
+pub trait StatePatcher {
+    async fn patch_state<U>(
+        &self,
+        contract_id: AccountId,
+        key: String,
+        value: &U,
+    ) -> anyhow::Result<()>
+    where
+        U: BorshSerialize + Send + Sync;
+}
+
+#[async_trait]
+impl<T> StatePatcher for T
+where
+    T: AllowStatePatching + NetworkClient + Send + Sync,
+{
```

Comment:
I feel like there should be a cleaner way to have this without exposing the AllowStatePatching trait. What you're trying to do here is just have a sealed trait impl, and you should be able to keep the AllowStatePatching trait private. I've taken a quick stab at this. This might not be the best alternative, but let me know your thoughts:

```
diff --git a/examples/src/spooning.rs b/examples/src/spooning.rs
index eff2653..20b059a 100644
--- a/examples/src/spooning.rs
+++ b/examples/src/spooning.rs
@@ -96,7 +96,11 @@ async fn main() -> anyhow::Result<()> {
 
     // Patch our testnet STATE into our local sandbox:
     worker
-        .patch_state(sandbox_contract.id(), "STATE".to_string(), &status_msg)
+        .patch_state(
+            sandbox_contract.id(),
+            "STATE".to_string(),
+            status_msg.try_to_vec().unwrap(),
+        )
         .await?;
 
     // Now grab the state to see that it has indeed been patched:
diff --git a/workspaces/src/network/mod.rs b/workspaces/src/network/mod.rs
index 30bcf02..21a734a 100644
--- a/workspaces/src/network/mod.rs
+++ b/workspaces/src/network/mod.rs
@@ -10,7 +10,6 @@ use async_trait::async_trait;
 
 use near_crypto::{InMemorySigner, KeyType, Signer};
 use near_jsonrpc_client::methods::sandbox_patch_state::RpcSandboxPatchStateRequest;
-use near_primitives::borsh::BorshSerialize;
 use near_primitives::state_record::StateRecord;
 use near_primitives::types::AccountId;
 
@@ -111,35 +110,31 @@ where
     }
 }
 
-pub trait AllowStatePatching {}
+pub(crate) mod sealed {
+    pub trait AllowStatePatching {}
+}
 
 #[async_trait]
-pub trait StatePatcher {
-    async fn patch_state<U>(
+pub trait StatePatcher: sealed::AllowStatePatching {
+    async fn patch_state(
         &self,
         contract_id: AccountId,
         key: String,
-        value: &U,
-    ) -> anyhow::Result<()>
-    where
-        U: BorshSerialize + Send + Sync;
+        value: Vec<u8>,
+    ) -> anyhow::Result<()>;
 }
 
 #[async_trait]
 impl<T> StatePatcher for T
 where
-    T: AllowStatePatching + NetworkClient + Send + Sync,
+    T: sealed::AllowStatePatching + NetworkClient + Send + Sync,
 {
-    async fn patch_state<U>(
+    async fn patch_state(
         &self,
         contract_id: AccountId,
         key: String,
-        value: &U,
-    ) -> anyhow::Result<()>
-    where
-        U: BorshSerialize + Send + Sync,
-    {
-        let value = U::try_to_vec(value).unwrap();
+        value: Vec<u8>,
+    ) -> anyhow::Result<()> {
         let state = StateRecord::Data {
             account_id: contract_id,
             data_key: key.into(),
diff --git a/workspaces/src/network/sandbox.rs b/workspaces/src/network/sandbox.rs
index 0e13426..6f2347b 100644
--- a/workspaces/src/network/sandbox.rs
+++ b/workspaces/src/network/sandbox.rs
@@ -9,7 +9,7 @@ use near_crypto::{InMemorySigner, Signer};
 use near_primitives::types::{AccountId, Balance};
 
 use super::{
-    Account, AllowDevAccountCreation, AllowStatePatching, CallExecution, Contract, NetworkClient,
+    Account, AllowDevAccountCreation, sealed, CallExecution, Contract, NetworkClient,
     NetworkInfo, TopLevelAccountCreator,
 };
 use crate::network::server::SandboxServer;
@@ -48,7 +48,7 @@ impl Sandbox {
     }
 }
 
-impl AllowStatePatching for Sandbox {}
+impl sealed::AllowStatePatching for Sandbox {}
 
 impl AllowDevAccountCreation for Sandbox {}
 
diff --git a/workspaces/src/prelude.rs b/workspaces/src/prelude.rs
index a499862..3f38e55 100644
--- a/workspaces/src/prelude.rs
+++ b/workspaces/src/prelude.rs
@@ -1,4 +1,4 @@
 pub use crate::network::{
-    AllowDevAccountCreation, AllowStatePatching, DevAccountDeployer, DevNetwork, Network,
+    AllowDevAccountCreation, DevAccountDeployer, DevNetwork, Network,
     StatePatcher, TopLevelAccountCreator,
 };
diff --git a/workspaces/src/worker/impls.rs b/workspaces/src/worker/impls.rs
index df803cb..a4cbac1 100644
--- a/workspaces/src/worker/impls.rs
+++ b/workspaces/src/worker/impls.rs
@@ -3,12 +3,11 @@ use std::path::Path;
 
 use async_trait::async_trait;
 use near_crypto::InMemorySigner;
-use near_primitives::borsh::BorshSerialize;
 use near_primitives::types::{AccountId, Balance, FunctionArgs, StoreKey};
 
 use crate::network::{
-    Account, AllowDevAccountCreation, CallExecution, CallExecutionDetails, Contract, NetworkClient,
-    NetworkInfo, StatePatcher, TopLevelAccountCreator,
+    sealed, Account, AllowDevAccountCreation, CallExecution, CallExecutionDetails, Contract,
+    NetworkClient, NetworkInfo, StatePatcher, TopLevelAccountCreator,
 };
 use crate::rpc::client::Client;
 use crate::worker::Worker;
@@ -74,20 +73,19 @@ where
     }
 }
 
+impl<T> sealed::AllowStatePatching for Worker<T> where T: sealed::AllowStatePatching {}
+
 #[async_trait]
 impl<T> StatePatcher for Worker<T>
 where
     T: StatePatcher + Send + Sync,
 {
-    async fn patch_state<U>(
+    async fn patch_state(
         &self,
         contract_id: AccountId,
         key: String,
-        value: &U,
-    ) -> anyhow::Result<()>
-    where
-        U: BorshSerialize + Send + Sync,
-    {
+        value: Vec<u8>,
+    ) -> anyhow::Result<()> {
         self.workspace.patch_state(contract_id, key, value).await
     }
 }
diff --git a/workspaces/tests/patch_state.rs b/workspaces/tests/patch_state.rs
index 4f1c513..49a6af8 100644
--- a/workspaces/tests/patch_state.rs
+++ b/workspaces/tests/patch_state.rs
@@ -72,7 +72,11 @@ async fn test_patch_state() -> anyhow::Result<()> {
     });
 
     worker
-        .patch_state(contract_id.clone(), "STATE".to_string(), &status_msg)
+        .patch_state(
+            contract_id.clone(),
+            "STATE".to_string(),
+            status_msg.try_to_vec().unwrap(),
+        )
         .await?;
 
     let result = worker

```

I also applied the change to Vec<u8> instead of being generic on borsh. This pattern can also be applied to the `AllowDevCreation` trait.

Reply:
The pattern I was going for here was to allow people to have access to the `AllowStatePatching` trait so that if people wanted to implement their own network, they would just `impl AllowStatePatching for CustomNetwork` and get `StatePatcher` for free, otherwise they impl their own custom `StatePatcher` if they needed. I'm not sure if what I did was the best way to convey that functionality though. Let me know if there's a better way to do that or if we want to go for something simpler than this.

Also, I did clean up the trait to take in `Vec<u8>` instead of `T: BorshSerialize`.

Reply:
hmm, yeah I suppose we want to expose the ability to configure a custom network, just seems excessive with the traits

Reply:
I don't know what the alternative would be though. Seems like this is as much as I can reduce it down to: as of now there's three traits in the prelude and two traits (`Network`, `DevNetwork`) we can directly import. I feel like those are all the essentials, and I've left some of the internal ones like `Allow*` to be exposed later